File: src/cli/xml2abx.cc
#include <cstring>
#include <fstream>
#include <iostream>
#include <sstream>

#include "abx.hpp"

#ifdef _WIN32
#include <io.h>
#define isatty _isatty
#define fileno _fileno
#else
#include <unistd.h>
#endif

void print_usage() {
    std::cerr << "usage: xml2abx [-i] [--collapse-whitespaces] input [output]\n\n"
              << "Converts between human-readable XML and Android Binary XML.\n\n"
              << "When invoked with the '-i' argument, the output of a "
                 "successful conversion\n"
              << "will overwrite the original input file. Input can be '-' to "
                 "use stdin, and\n"
              << "output can be '-' to use stdout.\n\n"
              << "Options:\n"
              << "  -i                     Overwrite input file with output\n"
              << "  --collapse-whitespaces Skip whitespace-only text nodes\n";
}

int main(int argc, char* argv[]) {
    bool in_place = false;
    bool collapse_whitespaces = false;
    std::string input_path;
    std::string output_path;

    int arg_idx = 1;
    while (arg_idx < argc) {
        if (strcmp(argv[arg_idx], "-i") == 0) {
            in_place = true;
            arg_idx++;
        } else if (strcmp(argv[arg_idx], "--collapse-whitespaces") == 0) {
            collapse_whitespaces = true;
            arg_idx++;
        } else if (strcmp(argv[arg_idx], "-h") == 0 || strcmp(argv[arg_idx], "--help") == 0) {
            print_usage();
            return 0;
        } else {
            break;
        }
    }

    if (arg_idx >= argc) {
        if (!isatty(fileno(stdin))) {
            input_path = "-";
            output_path = "-";
        } else {
            std::cerr << "Error: Missing input file\n\n";
            print_usage();
            return 1;
        }
    } else {
        input_path = argv[arg_idx++];

        if (in_place) {
            if (input_path == "-") {
                std::cerr << "Error: Cannot use -i flag with stdin\n";
                return 1;
            }
            if (arg_idx < argc) {
                std::cerr << "Error: Cannot specify output file with -i flag\n";
                return 1;
            }
            output_path = input_path;
        } else {
            if (arg_idx < argc) {
                output_path = argv[arg_idx++];
            } else {
                output_path = "-";
            }
        }
    }

    try {
        libabx::XmlToAbxOptions options;
        options.collapse_whitespaces = collapse_whitespaces;

        options.warning_callback = [](const std::string& category, const std::string& message) {
            std::cerr << "Warning [" << category << "]: " << message << "\n";
        };

        if (in_place) {
            std::ostringstream memory_output;

            libabx::convertXmlFileToAbx(input_path, memory_output, options);

            std::ofstream output_file(output_path, std::ios::binary);
            if (!output_file) {
                std::cerr << "Error: Cannot open output file: " << output_path << "\n";
                return 1;
            }
            output_file << memory_output.str();
            output_file.close();
        } else {
            std::ofstream output_file;
            std::ostream* output_stream = nullptr;

            if (output_path == "-") {
                output_stream = &std::cout;
            } else {
                output_file.open(output_path, std::ios::binary);
                if (!output_file) {
                    std::cerr << "Error: Cannot open output file: " << output_path << "\n";
                    return 1;
                }
                output_stream = &output_file;
            }

            if (input_path == "-") {
                std::ostringstream buffer;
                buffer << std::cin.rdbuf();
                libabx::convertXmlStringToAbx(buffer.str(), *output_stream, options);
            } else {
                libabx::convertXmlFileToAbx(input_path, *output_stream, options);
            }

            output_file.close();
        }

        return 0;

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
}

File: src/cli/xml2abx.rs
use android_xml_converter::{Options, convert_xml_string_to_buffer};
use std::fs;
use std::io::{self, Read, Write};
use std::process;

struct Args {
    input: String,
    output: Option<String>,
    in_place: bool,
    collapse_whitespaces: bool,
}

fn print_help() {
    eprintln!("Converts human-readable XML to Android Binary XML (ABX).");
    eprintln!();
    eprintln!("Usage: xml2abx [OPTIONS] <input> [output]");
    eprintln!();
    eprintln!("Arguments:");
    eprintln!("  <input>   Input file path (use '-' for stdin)");
    eprintln!("  [output]  Output file path (use '-' for stdout, defaults to stdout)");
    eprintln!();
    eprintln!("Options:");
    eprintln!("  -i, --in-place              Overwrite input file with converted output");
    eprintln!("  -c, --collapse-whitespaces  Remove unnecessary whitespace");
    eprintln!("  -h, --help                  Print this help message");
}

fn parse_args() -> Option<Args> {
    let args: Vec<String> = std::env::args().collect();
    let mut input = None;
    let mut output = None;
    let mut in_place = false;
    let mut collapse_whitespaces = false;
    let mut i = 1;

    while i < args.len() {
        match args[i].as_str() {
            "-h" | "--help" => {
                print_help();
                return None;
            }
            "-i" | "--in-place" => {
                in_place = true;
            }
            "-c" | "--collapse-whitespaces" => {
                collapse_whitespaces = true;
            }
            arg if arg.starts_with('-') && arg != "-" => {
                eprintln!("Error: Unknown option '{}'", arg);
                print_help();
                process::exit(1);
            }
            arg => {
                if input.is_none() {
                    input = Some(arg.to_string());
                } else if output.is_none() {
                    output = Some(arg.to_string());
                } else {
                    eprintln!("Error: Too many arguments");
                    print_help();
                    process::exit(1);
                }
            }
        }
        i += 1;
    }

    if input.is_none() {
        eprintln!("Error: Missing required argument <input>");
        print_help();
        process::exit(1);
    }

    Some(Args {
        input: input.unwrap(),
        output,
        in_place,
        collapse_whitespaces,
    })
}

fn read_input_string(input: &str) -> io::Result<String> {
    if input == "-" {
        let mut buffer = String::new();
        io::stdin().read_to_string(&mut buffer)?;
        Ok(buffer)
    } else {
        fs::read_to_string(input)
    }
}

fn write_output_xml2abx(output: Option<&str>, content: &[u8]) -> io::Result<()> {
    match output {
        Some(path) if path == "-" => io::stdout().write_all(content),
        None => io::stdout().write_all(content),
        Some(path) => fs::write(path, content).map(|_| ()),
    }
}

fn main() {
    let args = match parse_args() {
        Some(a) => a,
        None => return,
    };

    if args.in_place && args.input == "-" {
        eprintln!("Error: Cannot use --in-place with stdin");
        process::exit(1);
    }

    if args.in_place && args.output.is_some() {
        eprintln!("Error: Cannot specify output file with --in-place");
        process::exit(1);
    }

    let xml_content = match read_input_string(&args.input) {
        Ok(content) => content,
        Err(e) => {
            eprintln!("Error reading input: {}", e);
            process::exit(1);
        }
    };

    let options = Options::new().collapse_whitespaces(args.collapse_whitespaces);
    let abx_buffer = match convert_xml_string_to_buffer(&xml_content, Some(options)) {
        Ok(buffer) => buffer,
        Err(e) => {
            eprintln!("Error converting XML to ABX: {}", e);
            process::exit(1);
        }
    };

    let output_path = if args.in_place {
        Some(args.input.as_str())
    } else {
        args.output.as_deref()
    };

    if let Err(e) = write_output_xml2abx(output_path, &abx_buffer) {
        eprintln!("Error writing output: {}", e);
        process::exit(1);
    }
}
